# SVM

### Imports

In [1]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
from joblib import dump, load

In [2]:
from sklearn.svm import SVC
from sklearn.experimental import enable_halving_search_cv
from sklearn.model_selection import HalvingGridSearchCV

In [3]:
from preprocessing import reemplazarNulls,reemplazarCategoricas,reemplazarFechas,targetBooleano,regularizar

[###] All Done                                              


In [4]:
from utilities import score, df_feat, df_targ, train_val

### Codigo a correr

In [5]:
MAX_ITER = 10**6
MAX_JOBS = 5

**Run :** Entrenar Nuevamente o Cargar Entrenado

In [6]:
runLinearSimple = False
runLinear = False
runPoly = False
runRadial = False

**Save :** Guardar Modelo (pisa anterior)

In [7]:
saveLinear= True
saveLinearSimple= True
savePoly = True
saveRadial = True

### Preprocesamiento

In [8]:
targetBooleano(df_targ, inplace=True)
reemplazarNulls(df_feat , inplace=True)
reemplazarCategoricas(df_feat , inplace=True)
reemplazarFechas(df_feat , inplace=True)
df_allfeat = regularizar(df_feat , drop=0)
regularizar(df_feat , inplace=True)

id  direccion_viento_tarde  direccion_viento_temprano  \
0       0.495844                0.736537                   0.645996   
1       0.183431                0.317434                  -1.382279   
2      -0.113001               -0.730326                  -0.570969   
3      -1.678190                1.574745                   0.848823   
4       0.930508               -1.149430                  -1.179451   
...          ...                     ...                        ...   
102351 -0.264327               -0.101670                  -0.570969   
102352 -1.317418                0.946089                   1.457306   
102353 -1.594150                0.736537                   1.660133   
102354  0.293898               -1.149430                  -0.773796   
102355 -0.762408                1.574745                   1.457306   

        horas_de_sol  humedad_tarde  humedad_temprano  mm_lluvia_dia  \
0          -2.018030       1.277923          0.270523       7.914307   
1          -2.018030      -1.806334         -0.675093       0.417359   
2          -1.330311       1.326115         -1.147901      -0.281713   
3          -2.018030       1.952605          1.426276       0.320935   
4          -2.018030      -1.806334         -1.673244      -0.281713   
...              ...            ...               ...            ...   
102351     -0.298733       0.555051         -1.305504      -0.233501   
102352     -1.462565       0.651434          0.585729       0.345041   
102353     -2.018030       1.422498         -0.359888      -0.281713   
102354     -1.779974       1.326115         -0.202285      -0.281713   
102355      0.812198       0.217710         -0.202285      -0.281713   

        nubosidad_tarde  presion_atmosferica_tarde  \
0              0.916539                   1.269394   
1             -1.288420                  -2.406972   
2              1.284033                  -2.094694   
3              1.284033                   0.020281   
4             -0.920927                  -1.257220   
...                 ...                        ...   
102351         0.549046                  -0.490719   
102352         0.916539                   0.318365   
102353         0.916539                  -0.064885   
102354         1.284033                  -2.052110   
102355        -0.920927                   0.588060   

        rafaga_viento_max_direccion  rafaga_viento_max_velocidad  \
0                          0.802215                     0.589120   
1                          0.393094                     1.544505   
2                         -0.834272                     1.250541   
3                         -1.447954                     1.030067   
4                         -0.834272                     0.736103   
...                             ...                          ...   
102351                    -0.834272                     2.352908   
102352                     1.211337                     0.589120   
102353                    -1.447954                    -0.366265   
102354                    -0.629711                     1.177050   
102355                     1.415898                     0.589120   

        velocidad_viendo_tarde  
0                     0.834079  
1                     1.402051  
2                     1.629239  
3                     1.402051  
4                     0.038918  
...                        ...  
102351                2.083617  
102352                1.402051  
102353               -1.892186  
102354                0.606890  
102355                1.288456  

[102356 rows x 12 columns]

In [9]:
df_feat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 102356 entries, 0 to 102355
Data columns (total 12 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   id                           102356 non-null  float64
 1   direccion_viento_tarde       102356 non-null  float64
 2   direccion_viento_temprano    102356 non-null  float64
 3   horas_de_sol                 102356 non-null  float64
 4   humedad_tarde                102356 non-null  float64
 5   humedad_temprano             102356 non-null  float64
 6   mm_lluvia_dia                102356 non-null  float64
 7   nubosidad_tarde              102356 non-null  float64
 8   presion_atmosferica_tarde    102356 non-null  float64
 9   rafaga_viento_max_direccion  102356 non-null  float64
 10  rafaga_viento_max_velocidad  102356 non-null  float64
 11  velocidad_viendo_tarde       102356 non-null  float64
dtypes: float64(12)
memory usage: 9.4 MB


In [10]:
train_feat, val_feat, train_targ, val_targ = train_val( df_feat, df_targ )

## Entrenamiento

### Lineal

#### Simple

In [11]:
if (runLinearSimple):
    svm_simple_lineal = SVC(
        kernel='linear', C=0.1,
        max_iter=MAX_ITER, shrinking=False, verbose = 1,
    )
    svm_simple_lineal.fit( train_feat, train_targ.llovieron_hamburguesas_al_dia_siguiente )
    
    # Guardar
    if(saveLinearSimple):
        dump(svm_simple_lineal, 'models/SVM/svm_simple_lineal.sk') 
else:
    # Cargar
    svm_simple_lineal = load('models/SVM/svm_simple_lineal.sk')

[LibSVM]...............................................................*
optimization finished, #iter = 63625
obj = -3535.859021, rho = -1.284145
nSV = 35379, nBSV = 35346
Total nSV = 35379


In [12]:
%%time
linsim_pred = svm_simple_lineal.predict(val_feat)

CPU times: user 6.99 s, sys: 22.4 ms, total: 7.01 s
Wall time: 7.02 s


In [13]:
score(linsim_pred, val_targ.llovieron_hamburguesas_al_dia_siguiente)

ACCURACY: 0.8345056662758891


#### Optimizando Hiperparametros

In [14]:
base_arrange = np.arange(0.01, 0.09, 0.03)
hipr_arrange = np.concatenate((base_arrange,base_arrange*10,base_arrange*100),axis=None)
hipr_arrange

array([0.01, 0.04, 0.07, 0.1 , 0.4 , 0.7 , 1.  , 4.  , 7.  ])

In [15]:
if (runLinear):
    svm_lineal = SVC(
        kernel='linear',
        max_iter=MAX_ITER, shrinking=False,
    )
    params = {'C': hipr_arrange}
    
    svm_lineal_grid = HalvingGridSearchCV(
        svm_lineal,
        params,
        scoring='accuracy',
        n_jobs=MAX_JOBS,
        return_train_score=True,
        verbose=2,
    ).fit(train_feat, train_targ.llovieron_hamburguesas_al_dia_siguiente)
    
    # Guardar
    if(saveLinear):
        dump(svm_lineal_grid, 'models/SVM/svm_lineal_grid.sk') 
else:
    # Cargar
    svm_lineal_grid = load('models/SVM/svm_lineal_grid.sk')

n_iterations: 3
n_required_iterations: 3
n_possible_iterations: 3
min_resources_: 10235
max_resources_: 92120
aggressive_elimination: False
factor: 3
----------
iter: 0
n_candidates: 9
n_resources: 10235
Fitting 5 folds for each of 9 candidates, totalling 45 fits
----------
iter: 1
n_candidates: 3
n_resources: 30705
Fitting 5 folds for each of 3 candidates, totalling 15 fits
----------
iter: 2
n_candidates: 1
n_resources: 92115
Fitting 5 folds for each of 1 candidates, totalling 5 fits


/Users/nicozulaica/Documents/FIUBA/orga/venv/lib/python3.9/site-packages/sklearn/svm/_base.py:284: ConvergenceWarning: Solver terminated early (max_iter=1000000).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(
/Users/nicozulaica/Documents/FIUBA/orga/venv/lib/python3.9/site-packages/sklearn/svm/_base.py:284: ConvergenceWarning: Solver terminated early (max_iter=1000000).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(
/Users/nicozulaica/Documents/FIUBA/orga/venv/lib/python3.9/site-packages/sklearn/svm/_base.py:284: ConvergenceWarning: Solver terminated early (max_iter=1000000).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(
/Users/nicozulaica/Documents/FIUBA/orga/venv/lib/python3.9/site-packages/sklearn/svm/_base.py:284: ConvergenceWarning: Solver terminated early (max_iter=1000000).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warning

[CV] END .............................................C=0.01; total time=   1.4s
[CV] END .............................................C=0.04; total time=   1.4s
[CV] END ..............................C=0.06999999999999999; total time=   1.4s
[CV] END ..............................................C=0.1; total time=   1.6s
[CV] END ..............................................C=0.4; total time=   2.1s
[CV] END ..............................................C=0.7; total time=   3.7s
[CV] END ..............................................C=1.0; total time=   3.5s
[CV] END ..............................................C=4.0; total time=  13.6s
[CV] END ................................C=6.999999999999999; total time=  17.5s
[CV] END ..............................................C=1.0; total time=  39.3s
[CV] END ..............................................C=4.0; total time= 1.9min
[CV] END ................................C=6.999999999999999; total time= 3.2min
[CV] END ...................

/Users/nicozulaica/Documents/FIUBA/orga/venv/lib/python3.9/site-packages/sklearn/svm/_base.py:284: ConvergenceWarning: Solver terminated early (max_iter=1000000).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(


In [16]:
print(f"Best score: {svm_lineal_grid.best_score_}")
print(f"Best params {svm_lineal_grid.best_params_}")

Best score: 0.8392118547467838
Best params {'C': 6.999999999999999}


In [17]:
svm_lineal = svm_lineal_grid.best_estimator_

In [18]:
%%time
lin_pred = svm_lineal.predict(val_feat)

CPU times: user 7.01 s, sys: 19.6 ms, total: 7.03 s
Wall time: 7.04 s


In [23]:
score(lin_pred, val_targ.llovieron_hamburguesas_al_dia_siguiente)

ACCURACY: 0.8343102774521297


#### Otras Corridas

SVM Lineal , C=10, sin eliminar features poco significativos ni limitar iteraciones, t ~= 10hs \
El entranamiento fue removido debido a que tarda mucho y ni siquiera se esta optimizando el hiperparametro

In [24]:
svm_lineal_allfeat = load('models/SVM/svm_lineal_allfeat.sk')

/Users/nicozulaica/Documents/FIUBA/orga/venv/lib/python3.9/site-packages/sklearn/base.py:324: UserWarning: Trying to unpickle estimator SVC from version 0.23.2 when using version 1.0.1. This might lead to breaking code or invalid results. Use at your own risk. For more info please refer to:
https://scikit-learn.org/stable/modules/model_persistence.html#security-maintainability-limitations
  warnings.warn(


In [25]:
pred = svm_lineal_allfeat.predict(df_allfeat)
score(pred, df_targ.llovieron_hamburguesas_al_dia_siguiente)

/Users/nicozulaica/Documents/FIUBA/orga/venv/lib/python3.9/site-packages/sklearn/base.py:438: UserWarning: X has feature names, but SVC was fitted without feature names
  warnings.warn(


ACCURACY: 0.8404294814177967


### Poly

In [26]:
if (runPoly):
    svm_poly = SVC(
        kernel='poly',
        max_iter=MAX_ITER*10, shrinking=False,
    )
    params = {
        'C': hipr_arrange*10,
        'degree': np.arange(1,5,1),
        'gamma': hipr_arrange,
        'coef0': hipr_arrange*10,
    }
    
    svm_poly_grid = HalvingGridSearchCV(
        svm_poly,
        params,
        scoring='accuracy',
        n_jobs=MAX_JOBS,
        return_train_score=True,
        verbose=2,
    ).fit(train_feat, train_targ.llovieron_hamburguesas_al_dia_siguiente)
    
    # Guardar
    if(savePoly):
        dump(svm_poly_grid, 'models/SVM/svm_poly_grid.sk') 
else:
    # Cargar
    svm_poly_grid = load('models/SVM/svm_poly_grid.sk')

n_iterations: 8
n_required_iterations: 8
n_possible_iterations: 8
min_resources_: 42
max_resources_: 92120
aggressive_elimination: False
factor: 3
----------
iter: 0
n_candidates: 2916
n_resources: 42
Fitting 5 folds for each of 2916 candidates, totalling 14580 fits
[CV] END .............C=0.1, coef0=0.1, degree=1, gamma=0.01; total time=   0.0s
[CV] END .............C=0.1, coef0=0.1, degree=1, gamma=0.04; total time=   0.0s
[CV] END C=0.1, coef0=0.1, degree=1, gamma=0.06999999999999999; total time=   0.0s
[CV] END C=0.1, coef0=0.1, degree=1, gamma=0.06999999999999999; total time=   0.0s
[CV] END ..............C=0.1, coef0=0.1, degree=1, gamma=0.1; total time=   0.0s
[CV] END ..............C=0.1, coef0=0.1, degree=1, gamma=0.7; total time=   0.0s
[CV] END ..............C=0.1, coef0=0.1, degree=1, gamma=1.0; total time=   0.0s
[CV] END ..............C=0.1, coef0=0.1, degree=1, gamma=1.0; total time=   0.0s
[CV] END ..............C=0.1, coef0=0.1, degree=1, gamma=4.0; total time=   0.0s


In [27]:
print(f"Best score: {svm_poly_grid.best_score_}")
print(f"Best params {svm_poly_grid.best_params_}")

Best score: 0.842123026673925
Best params {'C': 0.7, 'coef0': 0.1, 'degree': 2, 'gamma': 0.06999999999999999}


In [28]:
svm_poly = svm_poly_grid.best_estimator_

In [29]:
%%time
pol_pred = svm_poly.predict(val_feat)

CPU times: user 7.56 s, sys: 21.6 ms, total: 7.58 s
Wall time: 7.59 s


In [30]:
score(pol_pred, val_targ.llovieron_hamburguesas_al_dia_siguiente)

ACCURACY: 0.8342125830402501


### Radial

In [31]:
if (runRadial):
    svm_radial = SVC(
        kernel='rbf',
        max_iter=MAX_ITER, shrinking=False,
    )
    params = {
        'C': hipr_arrange,
        'gamma': hipr_arrange,
    }
    
    svm_radial_grid = HalvingGridSearchCV(
        svm_radial,
        params,
        scoring='accuracy',
        n_jobs=MAX_JOBS,
        return_train_score=True,
        verbose=1,
    ).fit(train_feat, train_targ.llovieron_hamburguesas_al_dia_siguiente)
    
    # Guardar
    if(saveRadial):
        dump(svm_radial_grid, 'models/SVM/svm_radial_grid.sk') 
else:
    # Cargar
    svm_radial_grid = load('models/SVM/svm_radial_grid.sk')

n_iterations: 5
n_required_iterations: 5
n_possible_iterations: 5
min_resources_: 1137
max_resources_: 92120
aggressive_elimination: False
factor: 3
----------
iter: 0
n_candidates: 81
n_resources: 1137
Fitting 5 folds for each of 81 candidates, totalling 405 fits
----------
iter: 1
n_candidates: 27
n_resources: 3411
Fitting 5 folds for each of 27 candidates, totalling 135 fits
----------
iter: 2
n_candidates: 9
n_resources: 10233
Fitting 5 folds for each of 9 candidates, totalling 45 fits
----------
iter: 3
n_candidates: 3
n_resources: 30699
Fitting 5 folds for each of 3 candidates, totalling 15 fits
----------
iter: 4
n_candidates: 1
n_resources: 92097
Fitting 5 folds for each of 1 candidates, totalling 5 fits
[CV] END .............C=7.0, coef0=0.1, degree=3, gamma=0.04; total time=   0.0s
[CV] END .............C=7.0, coef0=0.1, degree=3, gamma=0.04; total time=   0.0s
[CV] END .............C=7.0, coef0=0.1, degree=3, gamma=0.04; total time=   0.0s
[CV] END .............C=7.0, coef0=

In [32]:
print(f"Best score: {svm_radial_grid.best_score_}")
print(f"Best params {svm_radial_grid.best_params_}")

Best score: 0.8470058092187415
Best params {'C': 4.0, 'gamma': 0.06999999999999999}


In [33]:
svm_radial = svm_radial_grid.best_estimator_

In [34]:
%%time
rad_pred = svm_radial.predict(val_feat)

CPU times: user 32 s, sys: 26.2 ms, total: 32 s
Wall time: 32 s


In [35]:
score(rad_pred, val_targ.llovieron_hamburguesas_al_dia_siguiente)

ACCURACY: 0.8389019148104728


## Resultados

#### Lineal

Best score: 0.8386942323663009 \
Best params {'C': 1.0}

ACCURACY: 0.8385927546992848

PRED TIME: 1min 20s

#### Polinomial

Best score: 0.842123026673925
Best params {'C': 0.7, 'coef0': 0.1, 'degree': 2, 'gamma': 0.06999999999999999}

ACCURACY: 0.83853413576146

PRED TIME: 1min 28s

#### Radial

Best score: 0.8438724989194568 \
Best params {'C': 0.4, 'gamma': 0.06999999999999999}

PRED TIME: 6min 4s

ACCURACY: 0.8459201219273906

### Coincidencia

In [36]:
(lin_pred == pol_pred).mean() *100

97.80187573270808

**TODO: ACA SE ME DESCHAVETO PQ EL POL YA NO ES UNA RECTA**

Corroboramos que los modelos lineal y polinomial, ademas de ser ambos lineales, \
tuvieron casi las mismas predicciones, debido a esto descartamos el polinomial

In [37]:
(lin_pred == rad_pred).mean() *100

96.47323173114498

Notamos que hay una pequeña diferencia entre las predicciones de cada uno, lo que podria significar cierto margen de mejora

In [39]:
( (val_targ.llovieron_hamburguesas_al_dia_siguiente == lin_pred) | (val_targ.llovieron_hamburguesas_al_dia_siguiente == rad_pred) ).mean() *100

85.42399374755763

In [41]:
( ( (val_targ.llovieron_hamburguesas_al_dia_siguiente == lin_pred) | (val_targ.llovieron_hamburguesas_al_dia_siguiente == rad_pred) ).mean() - (val_targ.llovieron_hamburguesas_al_dia_siguiente == rad_pred).mean() ) *100

1.5338022665103535

Un ensamble optimo de estos dos modelos podria tener como maximo un 1.13% de mejora \
No nos parece lo suficientemente significativo, ya que la mejora real seria significativamente menor \
y que seria mas efectivo entrenar modelos con el objetivo de ser ensamblados

### Conclusion

Descartamos el polinomial ya que es basicamente lineal sin ninguna mejora.

Entre el lineal y el radial, nos quedamos con ambos. \
Ya que el radial tiene metricas mejores pero es mas complejo y tarda mas en predecir

In [42]:
dump(svm_lineal, 'models/SVM/svm_lineal.sk')

['models/SVM/svm_lineal.sk']

In [43]:
dump(svm_radial, 'models/SVM/svm_radial.sk')

['models/SVM/svm_radial.sk']

In [44]:
dump(svm_poly, 'models/SVM/svm_poly.sk')

['models/SVM/svm_poly.sk']